### PR TITLE
[TIMOB-25572] Android: Provide Toolbar customization through DrawerLayout

### PR DESCRIPTION
**Pull request by ypbnv**
**JIRA:** https://jira.appcelerator.org/browse/TIMOB-25572

**Description:**
Add `customToolbar` creationOnly property to DrawerLayout which expects an instance of Ti.UI.Toolbar.
It would replace the embedded Toolbar and provide access to a greater customization.

**Test case:**
app.js
```
var win = Ti.UI.createWindow({theme: "Theme.AppCompat.NoTitleBar"});
    toolbar = Ti.UI.createToolbar({titleTextColor: 'red',
                                   backgroundColor: 'cyan'}),
    leftView = Ti.UI.createView({backgroundColor:'white'}),
    centerView = Ti.UI.createView({backgroundColor:'gray'}),
    drawer = Ti.UI.Android.createDrawerLayout({ toolbarEnabled: false,
                                                toolbar: toolbar,
                                                leftView: leftView,
                                                centerView: centerView }),
    button = Ti.UI.createButton({height:'100%', title: 'Toolbar'});

button.addEventListener('click', function() {
    drawer.toolbarEnabled = !drawer.getToolbarEnabled();
});
centerView.add(button);
win.add(drawer);
win.open();
```

_Documentation and possibly unit test are on the way._

### CODE REVIEW

File: android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUIDrawerLayout.java [hunk 1]
```diff
@@ -36,6 +37,7 @@
 import android.widget.RelativeLayout;
 
 import java.lang.reflect.Field;
+import java.util.HashMap;
```
Comment (garymathews) [inline android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUIDrawerLayout.java]: Is this used?

File: android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUIDrawerLayout.java [hunk 2]
```diff
@@ -414,6 +420,16 @@ public void processProperties(KrollDict d) {
 			toolbarEnabled = TiConvert.toBoolean(d.get(TiC.PROPERTY_TOOLBAR_ENABLED));
 			setToolbarVisible(toolbarEnabled);
 		}
+		if (d.containsKey(TiC.PROPERTY_CUSTOM_TOOLBAR)) {
+			if (!toolbarEnabled) {
+				if (((AppCompatActivity) proxy.getActivity()).getSupportActionBar() == null && (proxy.getActivity()).getActionBar() == null) {
```
Comment (garymathews) [inline android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUIDrawerLayout.java]: Maybe define `activity` to improve readability:
```Java
AppCompatActivity activity = (AppCompatActivity) proxy.getActivity();

if (activity.getSupportActionBar() == null && activity.getActionBar() == null) {
	View customToolbarInstance = ((ToolbarProxy) d.get(TiC.PROPERTY_CUSTOM_TOOLBAR)).getToolbarInstance();
	activity.setSupportActionBar((Toolbar) customToolbarInstance);
	RelativeLayout.LayoutParams layoutParams = new RelativeLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
	((RelativeLayout) layout.findViewById(id_drawer_layout_container)).addView(customToolbarInstance, layoutParams);
}
```

**Review by garymathews — CHANGES_REQUESTED** [review]
I think it would be better to have `toolbarEnabled == true` to enable either custom or default toolbar.

- If `toolbarEnabled == true` enable default toolbar unless `customToolbar` has been defined, then enable that instead.
- Minor, but maybe rename `customToolbar` into `toolbar`?

**Comment by ypbnv** [comment]
@garymathews Updated code and test case. I noticed that setting `toolbarEnabled=true` and using a theme with default ActionBar results in double toolbars. Would we want to prevent this in the Android side -  for instance ignoring the property when we have a default ActionBar?

**Review by garymathews — COMMENTED** [review]
@ypbnv Yes we would want to prevent that; could you update the PR for that scenario?

**Comment by ypbnv** [comment]
@garymathews Updated the PR.

**Comment by garymathews** [comment]
@ypbnv Looks good, docs?

**Comment by ypbnv** [comment]
@garymathews Updated PR. BTW do we support DrawerLayout through Alloy? It would be nice to add this there as well if we do.

**Comment by lokeshchdhry** [comment]
FR Passed.

The `toolBar` property for `DrawerLayout` shows the toolbar & the `toolbarEnabled` property works as expected.

Studio Ver:     5.0.0.201712081732
SDK Ver:        7.1.0 local build
OS Ver:         10.12.3
Xcode Ver:      Xcode 8.3.3
Appc NPM:       4.2.11
Appc CLI:       7.0.0
Daemon Ver:     1.0.0
Ti CLI Ver:     5.0.14
Alloy Ver:      1.10.10
Node Ver:       8.9.1
Java Ver:       1.8.0_101
Devices:        ⇨ google Nexus 5 --- Android 6.0.1
                ⇨ google Pixel --- Android 7.1.1

**Comment by lokeshchdhry** [comment]
@ypbnv , Can you please resolve the conflicts.

**Comment by lokeshchdhry** [comment]
@ypbnv , Can you please resolve the conflicts so that the PR can be merged.

**Comment by lokeshchdhry** [comment]
@ypbnv , can you please look at the unit test failures.

**Comment by TylerZubke** [comment]
Is the hamburger icon supposed to work correctly if I use a toolbar like this? I'm using a toolbar because I need the drawer to open above the actionbar (and as far as I know this is the only way), but the arrow always shows regardless of whether the drawer is open or not. Here's my app.js

`var win = Ti.UI.createWindow({theme: "Theme.AppCompat.NoTitleBar"});
toolbar = Ti.UI.createToolbar({titleTextColor: 'white',
    backgroundColor: 'black'}),
    leftView = Ti.UI.createView({backgroundColor:'white', opacity: 0.2}),
    centerView = Ti.UI.createView({backgroundColor:'gray'}),
    drawer = Ti.UI.Android.createDrawerLayout({ toolbarEnabled: true,
        toolbar: toolbar,
        leftView: leftView,
        centerView: centerView}),


    win.addEventListener("open", function() {
        win.activity.actionBar.displayHomeAsUp = true;
        win.activity.actionBar.onHomeIconItemSelected = function() {

            drawer.toggleLeft();
        }
    })


win.add(drawer);
win.open();
`

**Pull request merged**